vernac/vernacentries.ml

```diff
@@ -2717,16 +2735,30 @@ let interp ?(verbosely=true) ~st cmd =
     Vernacstate.invalidate_cache ();
     iraise exn
 
-let interp_qed_delayed_proof ~proof ~info ~st ?loc pe : Vernacstate.t =
+(* Interpreting a possibly delayed proof *)
+let interp_qed_delayed ~proof ~info ~st pe : Vernacstate.LemmaStack.t option =
   let stack = st.Vernacstate.lemmas in
   let stack = Option.cata (fun stack -> snd @@ Vernacstate.LemmaStack.pop stack) None stack in
+  let () = match pe with
+    | Admitted ->
+      save_lemma_admitted_delayed ~proof ~info
+    | Proved (_,idopt) ->
+      save_lemma_proved_delayed ~proof ~info ~idopt in
+  stack
+
+let interp_qed_delayed_control ~proof ~info ~st ?loc ~control pe : Vernacstate.t =
+  let vernac = CAst.make ?loc { control; attrs = []; expr = VernacEndProof pe } in
+  let stack =
+    List.fold_right (fun flag fn -> interp_control_flag ~vernac flag fn)
+      control
+      (fun ~st -> interp_qed_delayed ~proof ~info ~st pe)
+      ~st in
+  Vernacstate.Proof_global.set stack [@ocaml.warning "-3"];
+  Vernacstate.freeze_interp_state ~marshallable:false
```

Comment:
These last 2 lines didn't use to be there AFAICT, why the change?

Reply:
I think this was a bug in the previous implementation. In fact even the current one seems wrong, right? An `unfreeze` seems to be missing.

Maybe we should refactor the code so interp is shared.

The thing is that I'm quite close to remove the `interp_delayed` stuff in a separate branch reworking proof closing, but meanwhile we should try to make this sound.

Reply:
I think this was a bug in the previous implementation. In fact even the current one seems wrong, right? An `unfreeze` seems to be missing.

Maybe we should refactor the code so interp is shared.

The thing is that I'm quite close to remove the `interp_delayed` stuff in a separate branch reworking proof closing, but meanwhile we should try to make this sound.

Reply:
Ok, I've added a commit that tries to address this in a more systematic way.